### ISSUE TITLE
Ordering text formats no longer works

### ISSUE THREAD

**Issue by BWPanda** [backdrop/backdrop-issues#4407]
When trying to re-order the text formats on `/admin/config/content/formats`, they revert back to their old order on Save:

![Peek 2020-05-20 22-29](https://user-images.githubusercontent.com/2385329/82488959-835a1d80-9ae9-11ea-847e-bd9c8d06a409.gif)

This is on Backdrop v1.16.0. The row weights show the correct values, but in the wrong order.

---

PR by @quicksketch: https://github.com/backdrop/backdrop/pull/3210

**Comment by indigoxela** [comment]
Yep, confirmed.

First I checked if this is a form logic problem or a display problem - it's a _display-only_ issue in function filter_admin_overview(). The actual item weight and the display in the select list when adding/editing nodes are correct.

The problem arises here: https://github.com/backdrop/backdrop/blob/557824453eeec969800c2ff47db7bde4da036467/core/modules/filter/filter.admin.inc#L102

A simple `backdrop_sort()` won't do the trick here. We need to sort by the index of the form item. The "weight" is a form item, not a value, so not usable as a key for backdrop_sorg(). And additionally (actually before that) sort by "#status". Seems a bit tricky.

**Comment by jenlampton** [comment]
This is pretty serious since it has security implications, adding next milestone.

**Comment by indigoxela** [comment]
> This is pretty serious since it has security implications

I'm not sure if it actually has.
It's annoying and confusing for users, but the actual functionality doesn't seem to be affected. It's really only the display order on that overview page.

But that ordering thing is tricky. If someone has a smart idea how to solve this (I haven't), please post here - or even better, provide a PR.

**Comment by jenlampton** [comment]
Is https://github.com/backdrop/backdrop-issues/issues/4419 a duplicate of this issue? Would hat fix solve this problem also?

**Comment by indigoxela** [comment]
> Is #4419 a duplicate of this issue?

No, this issue is unrelated. Ordering isn't fixed by matching the parameters when calling filter_formats().

**Comment by quicksketch** [comment]
Filed a PR at https://github.com/backdrop/backdrop/pull/3210 that should fix this problem.

We finally ran into a situation where multiple sort orders would be useful, a topic we discussed in https://github.com/backdrop/backdrop-issues/issues/2149 but did not add at the time. Currently we can only sort ascending or descending, even if there are multiple keys. In this situation we want to sort on `#status` descending (1 before 0) and `#weight` ascending (-10, -9, 0, 1, 2, etc). But we can only sort one direction or the other in `backdrop_sort()`.

To side-step this I just instead added a large weight (+100) to disabled text formats to push them to the bottom, then sorted on weight only.
